xcm tool for NFTs

anshulWeb3 opened the pull request:
XCM tool for cross chain NFT
The Crosschain NFT Pallet is a unique initiative designed to facilitate the smooth movement of NFTs across various blockchain networks, utilizing the XCM protocol. Through the implementation of this pallet, individuals can effortlessly exchange NFTs between relay chains and parachains.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [x] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

keeganquigley commented:
Thanks @anshulWeb3 for the application.

- Can you integrate more technical details into the delivery table itself, such as the pallet functions?
- We only offer USDT payment on Statemint/Statemine. Please look at our [payment options](https://github.com/w3f/Grants-Program#pencil-process) for Ethereum.
- Are you able to provide the GitHub accounts of your team members, or otherwise provide proof of previous work?

Thanks!

anshulWeb3 commented:
Thankyou @keeganquigly for the review.
1. I have added details to the delivery table.
2. I have changed the payment address.
3. The names of team members are clickable and they have the github repos for the team members. Also, I have added the stack exchange profiles of my team members.
Kindly check.

keeganquigley commented:
Thanks for the changes @anshulWeb3 it's looking better now. I will mark the application as ready for review and ping the committee members internally. In the meantime, a few additional questions:

- Did this come out of a need to transfer your own NFTs, or otherwise what temperature checks have been done to let you know that this is needed in the community?
- Have you checked other [NFT bridging solutions](https://wiki.polkadot.network/docs/build-open-source) to see if some of these pallets have already been built? I seem to remember [RMRK](https://www.rmrk.app/) working on a substrate-to-substrate bridge but am not finding any updates with a quick search.

anshulWeb3 commented:
Thankyou @keeganquigley  for the questions.
1. To answer your first question, our team was researching on XCM and it's applications as we were looking into interoperability solutions. Cross chain NFT transfers interested us as NFT are a big part of functionality on blockchain and there are already many NFT projects that are live. Cross-chain support for NFT across the Substrate ecosystem could open a lot of possibilities in the future.
2. Yes, we looked into other bridging solutions. Work has been done for Fungible assets but we couldn't find anything relevant to our needs. Hence, we felt a need for this to exist. xTokens pallet exists why not xNFT. This project is also a part of our research on XCM and there is a lot more to come.

Noc2 reviewed (changes requested):
Thanks for the application. I have some questions: 
- Are you aware of [XCM V3](https://forum.polkadot.network/t/xcm-v3-is-coming-soon-what-new-possibilities-will-it-bring-to-polkadot-ecosystem/2313/1) and it's support for NFT transfers? 
- I assume that the biggest problem between different parchains and NFTs is standardization. How would you like to address this? What research/work has been done in regard to fungible assets that you could potentially leverage? Feel free to update the application accordingly. 
-

anshulWeb3 commented:
Thanks @Noc2  for reviewing the application.
1. Yes, we are aware of XCM v3 and we are building this xNFT pallet on top of the XCM v3 functionality for NFT transfers i.e the pallet consists of methods that combines the XCM functionality relevant for transferring NFT across chains. This pallet may be integrated with the substrate SDK as a module.
2. As of now our pallet supports the transfer of (only) NFT that are made using the NFT pallet provided by substrate. We plan to extend this functionality for custom NFTs in future. We have researched the xToken pallet (used for cross chain Fungible Asset transfers) currently used by parachains like Moonbeam and also have studied about Sovereign Account, TeleportAsset and ReserveTransferAsset. The architecture of the xNFT pallet will be somewhat similar to the xToken pallet.

takahser reviewed (commented):
@anshulWeb3 are you aware of [this issue](https://github.com/paritytech/substrate/issues/14349) and [this open PR](https://github.com/paritytech/substrate/pull/14395) that is related to that issue?

vstam1 commented:
Hi @anshulWeb3,

I think [this discussion](https://github.com/paritytech/xcm-format/pull/35) is also relevant for your tool.

anshulWeb3 commented:
Thankyou for the review @takahser . 
No, I just looked into this issue and the PR. Thankyou for bringing this into our notice. I believe that we will have to start our work once this PR is merged.

anshulWeb3 commented:
Hi @vstam1 , 
Thankyou for the reference.

takahser commented:
@anshulWeb3 thanks for your feedback.

> I believe that we will have to start our work once the PR you mentioned in your comment is merged.

I agree that it's a good idea to wait for the relevant changes on the NFT pallet to be implemented, but I'd generally be happy to support this effort once that's done. I think it's advisable to have a look at the [RFC](https://github.com/paritytech/xcm-format/pull/35) that @vstam1 has shared in [the comment above](https://github.com/w3f/Grants-Program/pull/1798#issuecomment-1598294428). Ideally, you'd get familiar with the changes that are currently being implemented and adapt your proposal accordingly.

In the meanwhile I'm going to put this application on hold, but feel free to further update it.

vstam1 commented:
I have to mention that the RFC is a proposal for a new instruction for XCMv4. It can still take some time before we release v4. The planning now is Q4 2023.

takahser commented:
@vstam1 do you think it'd be possible to parallelise the development of XCMv4 and this proposal? Maybe, once the relevant part of the XCMv4 API is fixed the grantee team could start implementing this?

vstam1 commented:
Unfortunately we cannot. XCM is versioned, and we cannot add another instruction to version 3.

Noc2 reviewed (approved):
Thanks for pinging me here. Given that's a level 1 grant, the application is already approved. But I'm happy to approve it as well and merge it.

keeganquigley commented:
Hi @anshulWeb3 did you want us to re-open the pull request? Are you still working on the milestone?

anshulWeb3 commented:
Hi @keeganquigley we are currently looking for solutions which align with "not changing the Frame NFT pallet" approach. Apart from that we are waiting for the required changes to get uptream.

keeganquigley commented:
Thanks @anshulWeb3 do you think these required changes will happen soon?

keeganquigley commented:
Hi @anshulWeb3 do you plan on submitting a delivery soon? If not, please consider submitting an amendment to extend the timeline. Thanks!

keeganquigley commented:
Hi @anshulWeb3 are you still working on this grant? I noticed there hasn't been much activity on the repos. Please let us know, otherwise after 2 weeks we will close the grant due to inactivity.

anshulWeb3 commented:
Hi @keeganquigley . Sorry for the delayed response. I was on a  Medical leave and the access to the account was with me only so any other team member was unable to respond in time. We've been engaged in research related to this matter and have explored various methods to access private fields such as ItemMetadataOf and CollectionMetadataOf within the FRAME NFT pallet, without altering the actual code. During this process, we've identified a solution and are currently in the implementation phase. We plan to provide you with a comprehensive update by next week. Thanks for your patience.

anshulWeb3 commented:
Hi @keeganquigley . We have successfully reverted all modifications made to the FRAME NFT pallet, including changes to structs such as CollectionDetails, ItemDetails, CollectionMetadata, and ItemMetadata. These structs now align with the standard FRAME NFT pallet.
However, we are currently addressing the final outstanding change, which involves making a Type of RuntimeCall in the Config of the FRAME NFT pallet to enable the calling of functions of the FRAME NFT pallet from the xnft pallet. We are actively working to resolve this issue and will provide updates as soon as it is resolved.
If you have any suggestions or insights regarding this matter, we are open to receiving them. Additionally, we will keep you informed about the resolution of the previous RuntimeCall issue as soon as it is resolved. Thank you for your understanding and cooperation.

semuelle commented:
Thanks for the update, @anshulWeb3. Sounds good to me. But could you submit an amendment to the grant application with an updated timeline? That makes it a lot easier for us to track project progress.

anshulWeb3 commented:
Hi @keeganquigley , @semuelle  We have successfully addressed the Runtime call challenge and are pleased to inform you that we will be delivering the xnft-pallet by the upcoming week.

anshulWeb3 commented:
Hi @keeganquigley . The development is done and we are  setting up the testing environment using Trappist. We expect to deliver by Wednesday. Thanks!

keeganquigley commented:
Great @anshulWeb3 sounds good thanks for the update!

anshulWeb3 commented:
Hello @keeganquigley . We are happy to announce that we have submitted the delivery of xNFT pallet. Looking forward to your response.  [Link to the PR](https://github.com/w3f/Grant-Milestone-Delivery/pull/1115).
Thanks.
